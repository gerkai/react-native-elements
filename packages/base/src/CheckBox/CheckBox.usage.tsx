import React from 'react';
import { CheckBox } from '..';
import { usage, Stack, info } from '@rneui/doc-gen';

info(
  'CheckBoxes allow users to complete tasks that involve making choices such as selecting options, or switching settings - On or Off. It provides a clear visual of either a true or false choice.'
);

usage(
  'Controlled CheckBox',
  "```tsx live\nfunction RNECheckBox() {const [open, setOpen] = React.useState(false);return (<Stack row align='center'><CheckBox checked={open} onPress={() => setOpen(!open)} /></Stack>);}\n ``` \n ### More examples",
  () => (
    <Stack row align="center">
      <CheckBox checked />
      <CheckBox checked checkedIcon="dot-circle-o" uncheckedIcon="circle-o" />
      <CheckBox
        center
        checked
        iconType="material"
        checkedIcon="clear"
        uncheckedIcon="add"
        checkedColor="red"
      />
    </Stack>
  )
);
