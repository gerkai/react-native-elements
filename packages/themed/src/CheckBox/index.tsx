import { withTheme } from '../config';
import {
  CheckBox,
  CheckBoxProps,
} from '@react-native-elements/base/dist/CheckBox/CheckBox';

export { CheckBox };
export type { CheckBoxProps };
export default withTheme(CheckBox, 'CheckBox');
