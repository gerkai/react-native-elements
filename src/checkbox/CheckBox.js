import React from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet,
  TouchableOpacity,
  View,
  Platform,
  Text as NativeText,
} from 'react-native';

import TextElement from '../text/Text';
import CheckBoxIcon from './CheckBoxIcon';
import { fonts, colors, ViewPropTypes } from '../config';

const CheckBox = props => {
  const {
    component,
    checked,
    iconRight,
    title,
    center,
    right,
    containerStyle,
    textStyle,
    onPress,
    onLongPress,
    checkedTitle,
    fontFamily,
    ...attributes
  } = props;

  const Component = component;

  return (
    <Component
      {...attributes}
      onLongPress={onLongPress}
      onPress={onPress}
      style={[
        styles.container,
        title && styles.containerHasTitle,
        containerStyle && containerStyle,
      ]}
    >
      <View
        style={[
          styles.wrapper,
          right && { justifyContent: 'flex-end' },
          center && { justifyContent: 'center' },
        ]}
      >
        {!iconRight && <CheckBoxIcon {...props} />}

        {React.isValidElement(title)
          ? title
          : title && (
              <TextElement
                style={[
                  styles.text,
                  textStyle && textStyle,
                  fontFamily && { fontFamily },
                ]}
              >
                {checked ? checkedTitle || title : title}
              </TextElement>
            )}

        {iconRight && <CheckBoxIcon {...props} />}
      </View>
    </Component>
  );
};

CheckBox.defaultProps = {
  checked: false,
  iconRight: false,
  right: false,
  center: false,
  checkedColor: colors.primary,
  uncheckedColor: '#bfbfbf',
  checkedIcon: 'check-square-o',
  uncheckedIcon: 'square-o',
  size: 24,
  component: TouchableOpacity,
};

CheckBox.propTypes = {
  ...CheckBoxIcon.propTypes,
  component: PropTypes.any,
  iconRight: PropTypes.bool,
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  center: PropTypes.bool,
  right: PropTypes.bool,
  containerStyle: ViewPropTypes.style,
  textStyle: NativeText.propTypes.style,
  onPress: PropTypes.func,
  onLongPress: PropTypes.func,
  checkedTitle: PropTypes.string,
  fontFamily: PropTypes.string,
};

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  container: {
    margin: 5,
    marginLeft: 10,
    marginRight: 10,
    padding: 10,
  },
  containerHasTitle: {
    borderWidth: 1,
    borderRadius: 3,
    backgroundColor: '#fafafa',
    borderColor: '#ededed',
  },
  text: {
    marginLeft: 10,
    marginRight: 10,
    color: colors.grey1,
    ...Platform.select({
      ios: {
        fontWeight: 'bold',
      },
      android: {
        ...fonts.android.bold,
      },
    }),
  },
});

export default CheckBox;
