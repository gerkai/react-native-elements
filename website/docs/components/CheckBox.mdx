---
id: checkbox
title: CheckBox
---

import { IconsStyle } from "@site/src/components/Docs_Icons";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";
import { BiInfoCircle } from "react-icons/bi";
import Usage from "../component_usage/CheckBox.mdx";
import Play from "@site/playground/CheckBox/checkbox.playground";

<IconsStyle />

CheckBoxes allow users to complete tasks that involve making choices such as selecting options, or switching settings - On or Off. It provides a clear visual of either a true or false choice.

## Usage

<div class="row inline-flex-center">
  <div class="col col--3">
    <h4>Import</h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`import { CheckBox } from '@rneui/themed';`}</CodeBlock>
  </div>
  <div class="col col--3">
    <h4>
      Theme Key{" "}
      <a href="../customizing#using-themeprovider">
        <BiInfoCircle />
      </a>
    </h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`CheckBox`}</CodeBlock>
  </div>
</div>

### Simple CheckBox

```tsx live showCode=false
function CheckBoxComponent() {
  const [state, setState] = React.useState(false);
  const toggleCheckbox = () => setState(!state);
  return (
    <Stack row align="center">
      <CheckBox checked={state} onPress={toggleCheckbox} />
      <CheckBox
        checked
        checkedIcon="dot-circle-o"
        uncheckedIcon="circle-o"
        onPress={toggleCheckbox}
      />
      <CheckBox
        center
        checked
        iconType="material"
        checkedIcon="clear"
        uncheckedIcon="add"
        checkedColor="red"
        onPress={toggleCheckbox}
      />
    </Stack>
  );
}
```

<Usage />

## Props

:::note
Includes all [CheckBoxIcon](checkboxicon#props), [View](https://reactnative.dev/docs/view#props) props.
:::

<div class='table-responsive'>

| Name                 | Type                                                                   | Default | Description                                     |
| -------------------- | ---------------------------------------------------------------------- | ------- | ----------------------------------------------- |
| `Component`          | React Component                                                        |         | Specify React Native component for main button. |
| `center`             | boolean                                                                | `false` | Aligns checkbox to center.                      |
| `checkedTitle`       | string                                                                 |         | Specify a custom checked message.               |
| `containerStyle`     | View Style                                                             |         | Style of main container.                        |
| `disabled`           | boolean                                                                | `false` | Disables user interaction.                      |
| `disabledStyle`      | View Style                                                             |         | Style of the checkbox container when disabled.  |
| `disabledTitleStyle` | Text Style                                                             |         | Style of the title when disabled.               |
| `fontFamily`         | string                                                                 |         | Specify different font family.                  |
| `iconRight`          | boolean                                                                | `false` | Moves icon to right of text.                    |
| `right`              | boolean                                                                | `false` | Aligns checkbox to right.                       |
| `textStyle`          | Text Style                                                             |         | Style of text.                                  |
| `title`              | `string` \| `ReactElement<{}, string` \| `JSXElementConstructor<any>>` |         | Title of checkbox.                              |
| `titleProps`         | TextProps                                                              | `{}`    | Additional props for the title Text component.  |
| `wrapperStyle`       | View Style                                                             |         | Style for the wrapper of checkbox.              |

</div>

## Playground

<Play />
