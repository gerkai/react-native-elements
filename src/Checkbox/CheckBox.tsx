import React from 'react';
import {
  StyleSheet,
  View,
  Platform,
  TextProps,
  TextStyle,
  ViewStyle,
  StyleProp,
  TouchableOpacityProps,
  TouchableOpacity,
} from 'react-native';
import TextElement from '../Text';
import { CheckBoxIcon, CheckBoxIconProps } from './components/CheckBoxIcon';
import { fonts } from '../config';
import { RneFunctionComponent } from '../helpers';

export type CheckBoxProps = TouchableOpacityProps &
  CheckBoxIconProps & {
    Component?: typeof React.Component;
    iconRight?: boolean;
    title?: string | React.ReactElement<{}>;
    titleProps?: TextProps;
    center?: boolean;
    right?: boolean;
    containerStyle?: StyleProp<ViewStyle>;
    wrapperStyle?: StyleProp<ViewStyle>;
    textStyle?: StyleProp<TextStyle>;
    checkedTitle?: string;
    fontFamily?: string;
  };

export const CheckBox: RneFunctionComponent<CheckBoxProps> = ({
  checked = false,
  Component = TouchableOpacity,
  iconRight = false,
  title,
  titleProps = {},
  center = false,
  right = false,
  containerStyle,
  wrapperStyle,
  textStyle,
  checkedTitle,
  fontFamily,
  theme,
  onPress,
  onLongPress,
  checkedColor = theme?.colors?.primary,
  ...rest
}) => {
  const accessibilityState = {
    checked: !!checked,
  };

  const iconProps = {
    checked,
    onLongPress,
    checkedColor,
    ...rest,
  };

  return (
    <Component
      accessibilityRole="checkbox"
      accessibilityState={accessibilityState}
      testID="checkbox"
      {...rest}
      onLongPress={onLongPress}
      onPress={onPress}
      style={StyleSheet.flatten([
        styles.container,
        title && styles.containerHasTitle,
        containerStyle && containerStyle,
      ])}
    >
      <View
        style={StyleSheet.flatten([
          styles.wrapper,
          right && { justifyContent: 'flex-end' },
          center && { justifyContent: 'center' },
          wrapperStyle && wrapperStyle,
        ])}
      >
        {!iconRight && (
          <CheckBoxIcon {...iconProps} checkedColor={checkedColor} />
        )}

        {React.isValidElement(title)
          ? title
          : title !== '' &&
            title && (
              <TextElement
                testID="checkboxTitle"
                style={
                  StyleSheet.flatten([
                    {
                      marginLeft: 10,
                      marginRight: 10,
                      color: theme?.colors?.grey1,
                      ...Platform.select({
                        android: {
                          ...fonts.android.bold,
                        },
                        default: {
                          fontWeight: 'bold',
                        },
                      }),
                    },
                    textStyle && textStyle,
                    fontFamily && { fontFamily },
                  ]) as any
                }
                {...titleProps}
              >
                {checked ? checkedTitle || title : title}
              </TextElement>
            )}

        {iconRight && (
          <CheckBoxIcon {...iconProps} checkedColor={checkedColor} />
        )}
      </View>
    </Component>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  container: {
    margin: 5,
    marginLeft: 10,
    marginRight: 10,
    padding: 10,
  },
  containerHasTitle: {
    borderWidth: 1,
    borderRadius: 3,
    backgroundColor: '#fafafa',
    borderColor: '#ededed',
  },
});

CheckBox.displayName = 'Checkbox';
