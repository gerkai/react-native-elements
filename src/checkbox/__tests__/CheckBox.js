import React from 'react';
import { Image } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import CheckBox from '../CheckBox';

describe('CheckBox Component', () => {
  it('should render without issues', () => {
    const component = shallow(<CheckBox />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should use TouchableOpacity as default component', () => {
    const component = shallow(<CheckBox />);

    expect(component.find('TouchableOpacity').length).toBe(1);
  });

  it('should allow to pass custom component', () => {
    const View = jest.fn();
    const component = shallow(<CheckBox component={View} />);

    expect(component.find('View').length).toBe(1);
  });

  it('should render title in Text', () => {
    const component = shallow(<CheckBox title="Custom Text" checked />);

    expect(component.props().children.props.children[1].props.children).toBe(
      'Custom Text'
    );
  });

  it('should render with icon and checked', () => {
    const component = shallow(
      <CheckBox
        iconType="font-awesome"
        checkedColor="red"
        containerStyle={{ backgroundColor: 'red' }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with icon and iconRight', () => {
    const component = shallow(
      <CheckBox
        iconType="font-awesome"
        iconRight
        uncheckedColor="blue"
        center
        right
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should allow custom checked Icon', () => {
    const component = shallow(
      <CheckBox
        checked={true}
        checkedIcon={
          <Image
            source={{ uri: 'https://image.ibb.co/jcY95H/checked.png' }}
            style={{ width: 30, height: 30 }}
          />
        }
        uncheckedIcon={
          <Image
            source={{ uri: 'https://image.ibb.co/fda0Cx/no_check.png' }}
            style={{ width: 30, height: 30 }}
          />
        }
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should allow custom checked Icon', () => {
    const component = shallow(
      <CheckBox
        checked={false}
        checkedIcon={
          <Image
            source={{ uri: 'https://image.ibb.co/jcY95H/checked.png' }}
            style={{ width: 30, height: 30 }}
          />
        }
        uncheckedIcon={
          <Image
            source={{ uri: 'https://image.ibb.co/fda0Cx/no_check.png' }}
            style={{ width: 30, height: 30 }}
          />
        }
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });
});
