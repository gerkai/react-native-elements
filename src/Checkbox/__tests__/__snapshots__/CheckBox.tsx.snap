// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox Component should allow custom checked Icon 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": true,
    }
  }
  checkedIcon={
    <Image
      source={
        Object {
          "uri": "https://image.ibb.co/jcY95H/checked.png",
        }
      }
      style={
        Object {
          "height": 30,
          "width": 30,
        }
      }
    />
  }
  style={
    Object {
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
  uncheckedIcon={
    <Image
      source={
        Object {
          "uri": "https://image.ibb.co/fda0Cx/no_check.png",
        }
      }
      style={
        Object {
          "height": 30,
          "width": 30,
        }
      }
    />
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={true}
      checkedColor="#2089dc"
      checkedIcon={
        <Image
          source={
            Object {
              "uri": "https://image.ibb.co/jcY95H/checked.png",
            }
          }
          style={
            Object {
              "height": 30,
              "width": 30,
            }
          }
        />
      }
      uncheckedIcon={
        <Image
          source={
            Object {
              "uri": "https://image.ibb.co/fda0Cx/no_check.png",
            }
          }
          style={
            Object {
              "height": 30,
              "width": 30,
            }
          }
        />
      }
    />
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should allow custom checked Icon when unchecked 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": false,
    }
  }
  checkedIcon={
    <Image
      source={
        Object {
          "uri": "https://image.ibb.co/jcY95H/checked.png",
        }
      }
      style={
        Object {
          "height": 30,
          "width": 30,
        }
      }
    />
  }
  style={
    Object {
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
  uncheckedIcon={
    <Image
      source={
        Object {
          "uri": "https://image.ibb.co/fda0Cx/no_check.png",
        }
      }
      style={
        Object {
          "height": 30,
          "width": 30,
        }
      }
    />
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={false}
      checkedColor="#2089dc"
      checkedIcon={
        <Image
          source={
            Object {
              "uri": "https://image.ibb.co/jcY95H/checked.png",
            }
          }
          style={
            Object {
              "height": 30,
              "width": 30,
            }
          }
        />
      }
      uncheckedIcon={
        <Image
          source={
            Object {
              "uri": "https://image.ibb.co/fda0Cx/no_check.png",
            }
          }
          style={
            Object {
              "height": 30,
              "width": 30,
            }
          }
        />
      }
    />
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should allow passing props to the title 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": true,
    }
  }
  style={
    Object {
      "backgroundColor": "#fafafa",
      "borderColor": "#ededed",
      "borderRadius": 3,
      "borderWidth": 1,
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={true}
      checkedColor="#2089dc"
    />
    <Themed.Text
      numberOfLines={2}
      style={
        Object {
          "color": "#43484d",
          "fontWeight": "bold",
          "marginLeft": 10,
          "marginRight": 10,
        }
      }
      testID="checkboxTitle"
    >
      Yea
    </Themed.Text>
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should render title in Text 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": true,
    }
  }
  style={
    Object {
      "backgroundColor": "#fafafa",
      "borderColor": "#ededed",
      "borderRadius": 3,
      "borderWidth": 1,
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={true}
      checkedColor="#2089dc"
    />
    <Themed.Text
      style={
        Object {
          "color": "#43484d",
          "fontWeight": "bold",
          "marginLeft": 10,
          "marginRight": 10,
        }
      }
      testID="checkboxTitle"
    >
      Custom Text
    </Themed.Text>
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should render with icon and checked 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": false,
    }
  }
  iconType="font-awesome"
  style={
    Object {
      "backgroundColor": "red",
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={false}
      checkedColor="red"
      iconType="font-awesome"
    />
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should render with icon and iconRight 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": false,
    }
  }
  iconType="font-awesome"
  style={
    Object {
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
  uncheckedColor="blue"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <CheckBoxIcon
      checked={false}
      checkedColor="#2089dc"
      iconType="font-awesome"
      uncheckedColor="blue"
    />
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should render with textStyle and fontFamily 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": true,
    }
  }
  style={
    Object {
      "backgroundColor": "#fafafa",
      "borderColor": "#ededed",
      "borderRadius": 3,
      "borderWidth": 1,
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={true}
      checkedColor="#2089dc"
    />
    <Themed.Text
      style={
        Object {
          "color": "red",
          "fontFamily": "serif",
          "fontWeight": "bold",
          "marginLeft": 10,
          "marginRight": 10,
        }
      }
      testID="checkboxTitle"
    >
      Custom Text
    </Themed.Text>
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should render with wrapperStyle 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": true,
    }
  }
  style={
    Object {
      "backgroundColor": "#fafafa",
      "borderColor": "#ededed",
      "borderRadius": 3,
      "borderWidth": 1,
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "red",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={true}
      checkedColor="#2089dc"
    />
    <Themed.Text
      style={
        Object {
          "color": "#43484d",
          "fontWeight": "bold",
          "marginLeft": 10,
          "marginRight": 10,
        }
      }
      testID="checkboxTitle"
    >
      Custom Text
    </Themed.Text>
  </View>
</ForwardRef>
`;

exports[`CheckBox Component should render without issues 1`] = `
<ForwardRef
  accessibilityRole="checkbox"
  accessibilityState={
    Object {
      "checked": false,
    }
  }
  style={
    Object {
      "margin": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "padding": 10,
    }
  }
  testID="checkbox"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <CheckBoxIcon
      checked={false}
      checkedColor="#2089dc"
    />
  </View>
</ForwardRef>
`;
